### ISSUE TITLE
The certificate that you're adding to HttpClientHandler.ClientCertificates only contains the public key.  The SSL connection can't happen without access to the private key.

### ISSUE THREAD

**Issue by MatthiasBauernberger** [dotnet/runtime#69199]
The certificate that you're adding to HttpClientHandler.ClientCertificates only contains the public key.  The SSL connection can't happen without access to the private key.

Using 'curl', specifying the public key on the command line makes curl look up the private key in the machine configuration.

There are 2 ways that a private key is specified in HttpClient.  You can read the entire client certificate from the system which will include both public and private key portions in an X509Certificate2 object.  Then add that to the HttpClientHandler.ClientCertificates.

The other way is to only add the public key portion (which is what you did).  But then HttpClient is supposed to look up the private key in the system if it is missing from the X509Certificate2 object.

This behavior of looking up the private key portion (if missing from the certificate) works on .NET Framework (Windows).  It currently does NOT work on .NET Core (Windows) since that was not implemented.  This is a current feature gap of .NET Core vs. .NET Framework.

I suspect that the private key lookup is also NOT implemented in the .NET Core Linux implementation. And that is why you are getting an error.

cc: @stephentoub @bartonjs

_Originally posted by @davidsh in https://github.com/dotnet/runtime/issues/23074#issuecomment-320523332_

**Comment by MatthiasBauernberger** [comment]
Hi - Is the difference between .NET Framework and .NET Core (these days .NET 6) still there. I have .NET Core webclient and tried to get the private key automatically added (like in the old days) by using the cer-file (public key) but it doesn't work. Currently I have to use a PKCS12 (p12 or pfx) file, to send a client certificate to the my server.  I don't like this approach because I have to have the password for the private key in my code. 

Thank you in advance for your response.
Best regards, Matthias

**Comment by bartonjs** [comment]
The "find the private key" logic exists, but requires that you have previously added the cert+key into the CurrentUser\My store.

```C#
// Once ever.
using (X509Store store = new X509Store(StoreName.My, StoreLocation.CurrentUser, OpenFlags.ReadWrite))
{
    using (X509Certificate2 cert = new X509Certificate2(pfx, pwd, X509KeyStorageFlags.PersistKeySet))
    {
        store.Add(cert);
    }
}
```

Now SslStream should be able to match against the store and find the cert.  (Alternatively, you could just find it in the store before calling SslStream).

**Comment by wfurt** [comment]
As @bartonjs mentioned the logic exist but does not come cheap and failures are difficult to diagnose. As you mentioned, this private key must be reachable - either via store of via pix load. I would strongly suggest to get the X509Certificate2 with private kay from the store and use application logic to log any errors.

**Comment by MatthiasBauernberger** [comment]
Thank you for your quick responses. So basically the way it was (.NET Framework) changed with .NET Core/.NET 6 and my options are as it outlined by @bartonjs or I'll go with the solution which I'm already having in place to use the pfx + password, which requires of course to set the necessary security (install pfx -> manage private keys -> by using certificate mmc for the security context (AppPoolIdentity) where the .NET 6 app run in). 

`
// My current code which works (installed pfx, correct AppPool-Settings, correct "manage private key" within certificate mmc)
 HttpClientHandler httpClientHandler = new();
            X509Certificate2 x509Certificate2 = new(pfxPath, Constants.Pkcs12Pwd);
            httpClientHandler .ClientCertificates.Add(x509Certificate2);
            using HttpClient httpClient = new(httpClientHandler );
...
`

**Bottom line:** Adding a client certificate to a C# Webrequest (or an HttpClient) by using the public key (cer file) of the installed certificate isn't available in .NET 6 as it was in .NET Framework. Right?

Best regards, 
Matthias

**Comment by wfurt** [comment]
The lookup does exist. I'm not sure if it is 100% same as .NET Framework. 
This is implementation for 6.0. For client, it will look in `StoreLocation.CurrentUser` 

https://github.com/dotnet/runtime/blob/dc4d2d34a39c89d2a9903dcff8bb2841d99412a2/src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs#L187-L191

**Comment by karelz** [comment]
@MatthiasBauernberger do you have repro for the problem? We believe it should work on 6.0+.

**Comment by karelz** [comment]
Triage: No response, closing. Feel free to reopen when there is additional info available. Thanks!
